File: app_examples/gpu-canvas/tests/browser.mjs
import assert from 'node:assert/strict';
import { mkdir } from 'node:fs/promises';

assert.equal(process.env.ARGUI_HIDDEN_DISPLAY, '1', 'Run with scripts/linux-hidden-display.sh');
assert.equal(process.env.DISPLAY, undefined, 'Do not inherit the desktop X11 display');
const imported = await import(process.env.PUPPETEER_MODULE ?? 'puppeteer');
const puppeteer = imported.puppeteer ?? imported.default;
const output = process.env.SCREENSHOT_DIR ?? 'target/gpu-canvas-browser';
await mkdir(output, { recursive: true });

const browser = await puppeteer.launch({
    executablePath: process.env.CHROME_PATH,
    headless: false,
    args: [
        '--ozone-platform=wayland',
        '--enable-unsafe-webgpu',
        '--ignore-gpu-blocklist',
        '--enable-features=Vulkan',
        '--use-angle=vulkan',
    ],
});

try {
    const page = await browser.newPage();
    const errors = [];
    page.on('pageerror', error => errors.push(String(error)));
    page.on('console', message => {
        if (message.type() === 'error' && !message.text().includes('simulated failure')) {
            errors.push(message.text());
        }
    });
    await page.setViewport({ width: 1280, height: 780, deviceScaleFactor: 1 });
    await page.evaluateOnNewDocument(() => {
        window.arguiRendererState = 'loading';
        window.arguiFrameCount = 0;
        window.addEventListener('argui:renderer-state', event => {
            window.arguiRendererState = event.detail.state;
        });
        const raf = window.requestAnimationFrame.bind(window);
        window.requestAnimationFrame = callback => raf(time => {
            window.arguiFrameCount++;
            callback(time);
        });
    });
    await page.goto(process.env.GPU_CANVAS_URL ?? 'http://127.0.0.1:8794/', {
        waitUntil: 'networkidle0',
    });
    await page.waitForFunction(() => window.arguiRendererState === 'ready');
    await page.waitForSelector('[role="img"][aria-label="Interactive GPU particle canvas"]');

    const capture = async name => {
        const png = await page.screenshot({ path: `${output}/${name}.png` });
        const populated = await page.evaluate(async base64 => {
            const image = new Image();
            image.src = `data:image/png;base64,${base64}`;
            await image.decode();
            const canvas = document.createElement('canvas');
            canvas.width = image.width;
            canvas.height = image.height;
            const context = canvas.getContext('2d');
            context.drawImage(image, 0, 0);
            const pixels = context.getImageData(0, 0, image.width, image.height).data;
            const colors = new Set();
            for (let index = 0; index < pixels.length; index += 64) {
                colors.add((pixels[index] << 16) | (pixels[index + 1] << 8) | pixels[index + 2]);
                if (colors.size > 48) return true;
            }
            return false;
        }, Buffer.from(png).toString('base64'));
        assert.ok(populated, `Blank GPU Canvas Lab capture: ${name}`);
    };
    const button = label => `button[aria-label="${label}"]`;
    const click = async label => {
        await page.waitForSelector(button(label));
        await page.$eval(button(label), element => element.click());
        await new Promise(resolve => setTimeout(resolve, 250));
    };
    const waitForAccessibleText = value => page.waitForFunction(
        expected => [...document.querySelectorAll('[aria-label]')]
            .some(element => element.getAttribute('aria-label')?.includes(expected)),
        {},
        value,
    );
    const canvas = await page.$('[role="img"][aria-label="Interactive GPU particle canvas"]');
    const rect = await canvas.boundingBox();
    assert.ok(rect && rect.width > 400 && rect.height > 300, 'Canvas viewport is visible');

    await capture('initial');
    await page.mouse.move(rect.x + rect.width * 0.55, rect.y + rect.height * 0.55);
    await page.mouse.down();
    await page.mouse.move(rect.x + rect.width * 0.68, rect.y + rect.height * 0.44, { steps: 8 });
    await page.mouse.up();
    await page.mouse.wheel({ deltaY: -240 });
    await page.waitForFunction(() => [...document.querySelectorAll('[aria-label]')]
        .map(element => element.getAttribute('aria-label'))
        .some(label => {
            const match = label?.match(/(\d+\.\d{2})×/);
            return match && Number.parseFloat(match[1]) > 1.05;
        }),
        { timeout: 10000 });
    await click('Zoom +');
    await capture('panned-zoomed');

    await click('Test recovery');
    await waitForAccessibleText('simulated failure');
    await capture('diagnostic-placeholder');
    await click('Recover canvas');
    await waitForAccessibleText('recovered');

    await click('Pause animation');
    await new Promise(resolve => setTimeout(resolve, 400));
    const before = await page.evaluate(() => window.arguiFrameCount);
    await new Promise(resolve => setTimeout(resolve, 700));
    const after = await page.evaluate(() => window.arguiFrameCount);
    assert.equal(after, before, 'Paused canvas requests no animation frames');

    await page.setViewport({ width: 980, height: 700, deviceScaleFactor: 2 });
    await click('Resume animation');
    await new Promise(resolve => setTimeout(resolve, 350));
    await capture('hidpi-resized-resumed');
    assert.deepEqual(errors, []);
    console.log(JSON.stringify({ screenshots: output, pausedFrames: after - before }, null, 2));
} finally {
    await browser.close();
}

File: app_examples/astra-editor/tests/browser.py
#!/usr/bin/env python3
"""Exercise Astra Editor's WebAssembly build in a private visible Chromium."""

import io
import os
from pathlib import Path
import time

from PIL import Image
from playwright.sync_api import sync_playwright


def contrasting(png: bytes) -> bool:
    """Return whether a PNG contains enough contrast to be a real rendered frame."""
    image = Image.open(io.BytesIO(png)).convert("RGB")
    return max(high - low for low, high in image.getextrema()) > 100


def main() -> None:
    """Open, interact with, resize, and capture the browser editor."""
    assert os.environ.get("ARGUI_HIDDEN_DISPLAY") == "1"
    assert "DISPLAY" not in os.environ
    output = Path(os.environ.get("SCREENSHOT_DIR", "target/astra-editor-browser"))
    output.mkdir(parents=True, exist_ok=True)
    url = os.environ.get(
        "ASTRA_EDITOR_URL", "http://127.0.0.1:8795/examples/astra-editor/"
    )
    errors: list[str] = []

    with sync_playwright() as playwright:
        browser = playwright.chromium.launch(
            executable_path=os.environ["CHROME_PATH"],
            headless=False,
            args=[
                "--ozone-platform=wayland",
                "--enable-unsafe-webgpu",
                "--ignore-gpu-blocklist",
                "--enable-features=Vulkan",
                "--use-angle=vulkan",
            ],
        )
        try:
            page = browser.new_page(viewport={"width": 1280, "height": 800})
            page.on("pageerror", lambda error: errors.append(str(error)))
            page.on(
                "console",
                lambda message: errors.append(message.text)
                if message.type == "error"
                else None,
            )
            page.goto(url, wait_until="networkidle", timeout=90_000)
            page.wait_for_function(
                "document.querySelector('#status')?.hidden === true", timeout=90_000
            )
            canvas = page.locator("canvas")
            canvas.wait_for(state="visible")
            bounds = canvas.bounding_box()
            assert bounds and bounds["width"] == 1280 and bounds["height"] == 800
            initial = page.screenshot(path=output / "desktop.png")
            assert contrasting(initial), "blank desktop WebAssembly capture"

            canvas.focus()
            page.keyboard.press("Control+b")
            time.sleep(0.35)
            explorer_hidden = page.screenshot(path=output / "explorer-hidden.png")
            assert explorer_hidden != initial, "Ctrl+B did not change the browser frame"
            page.keyboard.press("Control+b")
            page.keyboard.press("Control+Shift+f")
            page.get_by_role("dialog", name="Search across files").wait_for(
                state="visible"
            )
            page.keyboard.press("Escape")

            page.set_viewport_size({"width": 430, "height": 860})
            time.sleep(0.4)
            compact = page.screenshot(path=output / "compact.png")
            assert contrasting(compact), "blank compact WebAssembly capture"
            assert page.evaluate("document.documentElement.scrollWidth === innerWidth")
            assert errors == [], errors
        finally:
            browser.close()


if __name__ == "__main__":
    main()
